### ISSUE TITLE
Large resources corrupted with Core CLR on Linux

### ISSUE THREAD

**Issue by dougbu** [dotnet/roslyn#5052]
With Core CLR on Linux, resources larger than `4k` are corrupted in assemblies on disk and in memory. Looking at the relevant DNX code, I'm pretty sure this issue arises lower in the stack. But it be either a Core Fx or Roslyn issue. Starting here but please let me know if it looks more like a Core Fx problem.

For example the sample at [dougbu/LargeResources](https://github.com/dougbu/LargeResources) outputs the following normally:
```
Success for resources/a.txt.
Success for resources/b.txt.
Success for resources/c.txt.
Success for resources/d.txt.
```

With Core CLR on Linux, the output is instead:
```
Success for resources/a.txt.
Mismatch at index '4096' in resources/b.txt.
        File contains     '...bbbbbbbbbbbbbbbbbbbb'.
        Resource contains '...aaaaaaaaaaaa'.
Success for resources/c.txt.
Mismatch at index '4096' in resources/d.txt.
        File contains     '...d'.
        Resource contains '...'.
```

Resource corruption is also visible in `ILSpy` with an assembly generated on Linux i.e. `dnu pack` then transferred to Windows.

PS
- The [sample application](https://github.com/dougbu/LargeResources) must be executed in the directory containing the `project.json` file.
- Corrupted segments contain `null` bytes, making the resource output appear shorter in the comparison above.
- The corruption offset changes to 2047 if the resource files are UTF16-encoded (with a BOM). This implies the issue relates to the (lower) `byte`-oriented layers in the system.

**Comment by gafter** [comment]
We should investigate this problem in the 1.1 milestone and either fix the bug or dispatch it to the appropriate team.

**Comment by muratg** [comment]
We are getting customer reports on this issue as well.

**Comment by agocke** [comment]
This is a bug in CoreFX: https://github.com/dotnet/corefx/issues/3258

I'm closing the bug on Roslyn for now.

**Comment by dougbu** [comment]
Thanks for the investigation.

**Comment by stephentoub** [comment]
I'm not convinced it's a bug in FileStream.  The repro supplied assumes that Stream.Read always reads the number of requested bytes, which isn't the case; it can read and give back fewer than the number requested.  I don't know if it's the relevant Roslyn code, but I see code like https://github.com/dotnet/roslyn/blob/0382e3e3fc543fc483090bff3ab1eaae39dfb4d9/src/Compilers/Core/Portable/CvtRes.cs#L207-L209 that's similarly ignoring the return value of Read.

**Comment by agocke** [comment]
@stephentoub You are absolutely correct. It looks like we've made this error in multiple places, so we'll need to fix them up. Thanks for catching my error!

**Comment by stephentoub** [comment]
:smile: In addition to your fixing that, I've also submitted a PR to corefx to bring the Unix behavior of FileStream a bit more in line with that on Windows.  On Windows, after filling a read from the internal buffer, if we didn't get as much from the buffer as the user asked for, we'll make another request to the OS to fill the remainder.  My change modifies the Unix behavior to be similar.  However, it's still not guaranteed to be as much as the user requested (e.g. POSIX's read can return less than what is asked for, just as .NET's Read) and your fixes in Roslyn are still important.

**Comment by agocke** [comment]
Fixed by #5265
